auth: Use authmap for auth_required policies

jrajahalme opened the pull request:
Use the auth map (added in PR #24218) for auth_required policy instead of the CT map.

jrajahalme committed:
datapath: Use auth map for policy

Use the new auth map for auth_required policy instead of the CT map.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
Revert "bpf: Add CT map flag for auth required"

This reverts commit 1f7ed72e6b132202caaa085d6774bb62f58ca7b4.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
maps/auth: rename package to authmap

Rename pkg/maps/auth as pkg/maps/authmap to avoid requiring explicitly
importing it as "authmap" when also importing the pkg/auth.

Most of the other maps packages have the "map" suffix, so adds to this
consistency in package naming.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

squeed reviewed (approved):
I only reviewed my files; looks fine.

youngnick reviewed (approved):
LGTM for the Go bits.

joestringer commented on bpf/lib/host_firewall.h:
```diff
@@ -24,7 +25,8 @@ ipv6_host_policy_egress(struct __ctx_buff *ctx, __u32 src_id,
 	struct remote_endpoint_info *info;
 	struct ipv6_ct_tuple tuple = {};
-	__u32 dst_id = 0;
+	__u32 dst_id = 0; /* default to WORLD_ID instead? */
```
`lookup_ip6_remote_endpoint()` should find the `::/0` entry in ipcache anyway and fall back to world.

joestringer replied:
(Similar for the comment on the IPv4 version)

jrajahalme replied:
Thanks, dropped these comments.

joestringer commented on bpf/lib/host_firewall.h:
In principle I think that it's OK to make the same changes to host firewall as for endpoint firewalling, however in practice we must poke holes for the authentication traffic, otherwise if a user configures a policy that forces authentication for every connection outbound, it will create a 🐔🥚 failure when the agent attempts to initiate an authentication session from the node through the host firewall, which would then also be dropped, triggering another authentication attempt, and so on.

jrajahalme replied:
Good point. At first we should document that if host firewall is enabled then the host-to-host auth traffic must be explicitly allowed and not requiring authentication (to avoid the 🐔🥚 failure).

Opened issue https://github.com/cilium/cilium/issues/24542 to poke this hole automatically when both host firewall and Cilium mTLS are enabled.

joestringer commented on pkg/auth/authmap_authenticator.go:
```diff
@@ -0,0 +1,58 @@
+}
+
+// GetBoottime returns the kernel boot time.
+// We parse it from /proc/stat. GetBoottime() should be invoked only once.
```
nit: Maybe rather than asking for it to only be invoked once, this code could use `sync.Once` to force it?

jrajahalme replied:
Changed this to be called periodically to catch NTP time drifts and also the delta between monotonic and boottime clocks as Linux < 5.8 does not have bpf helper for `ktime_get_boot_ns()`.

joestringer commented on pkg/auth/authmap_authenticator.go:
```diff
@@ -0,0 +1,58 @@
+}
+
+func (r *authMapAuthenticator) markAuthenticated(result *authResult) error {
+	return r.authMap.Update(result.localIdentity, result.remoteIdentity, result.remoteNodeID, result.authType, uint64(result.expirationTime.Sub(r.bootTime).Nanoseconds()))
```
Why does this subtract boot time when inserting into the map?

jrajahalme replied:
At the time you reviewed this, the expiry time was thought to be in units returned by `ktime_get_boot_ns()`, i.e., nanoseconds since boot. That is why the boot time was subtracted.

Now, in the updated version we only use `ktime_get_ns()`, so we need to account for the suspend time with the help of userspace. As part of this change the expiry time is not relative to the Unix epoch rather than the system boot. This fixes the need to update the expiry times if there is any NTP time drift, as such drift effectively moves the system boot time around, leading to the need to do the subtraction again.

joestringer commented on bpf/lib/auth.h:
```diff
@@ -0,0 +1,34 @@
+		/* check that entry has not expired */
+		__u64 ktime = ktime_get_boot_ns();
+
+		if (ktime < auth->expiration)
```
Do we need to consider anything particular about time sources to ensure they're comparable between the userspace and datapath implementations?

jrajahalme replied:
This is explicitly addressed in the new version by addition of the `utime` (array) map.

joestringer commented on bpf/bpf_lxc.c:
```diff
@@ -419,16 +420,13 @@ static __always_inline int handle_ipv6_from_lxc(struct __ctx_buff *ctx, __u32 *d
-		}
+	if (verdict == DROP_POLICY_AUTH_REQUIRED && ct_status == CT_NEW) {
+		ct_state_new.src_sec_id = SECLABEL;
+		ct_create6(get_ct_map6(tuple), &CT_MAP_ANY6, tuple, ctx,
```
Do we still need to create connection tracking entries with the new approach?

joestringer replied:
Ah, at a glance it looks like these are now being removed by the last revert commit :+1:

joestringer commented on pkg/auth/manager.go:
```diff
@@ -52,7 +53,7 @@ type authResult struct {
 	localIdentity  identity.NumericIdentity
 	remoteIdentity identity.NumericIdentity
-	remoteHostIP   net.IP
+	remoteNodeID   uint16
```
This is only created & used locally to the agent right? (Just checking since I believe these are not synchronized across the cluster)

jrajahalme replied:
Yes, locally in the agent, and for remote nodes only (local node has NodeID `0`).

mhofstetter commented on pkg/auth/authmap_authenticator.go:
```diff
@@ -0,0 +1,58 @@
+	log.Infof("Boottime: %s", boottime.String())
+
+	return &authMapAuthenticator{
+		authMap:  authmap.AuthMap(),
```
it seems like the authmap has not yet been initialized when calling this. which results in setting the `authMap` to `nil` which potentially results in nil pointer exceptions when trying to access it (https://github.com/cilium/cilium/pull/24410#issuecomment-1479501843).

this is due to initialization order issues:
1. authmanager initialization (and therefore authmap_authenticator)
2. daemon initialization (which depends on the authmanager for registering it to the agent monitor)
3. daemon initializes the authmap (only after this point the called function `authmap.AuthMap()` returns something.

I already addressed this in https://github.com/cilium/cilium/pull/24406, by providing the authmap as hive cell with the proper initialization happening during the hive lifecycle hook.

mhofstetter replied:
@jrajahalme 

instead of cherry-picking the commit from the referenced PR and pulling in additional context into this PR - it might be worth to workaround this issue by simply initializing the authmap before accessing it within the authenticator by adding the following lines before calling `authmap.AuthMap()`.

```
	if err := authmap.InitAuthMap(option.Config.AuthMapEntries); err != nil {
		panic(err)
	}
```

jrajahalme replied:
Done :-)

jrajahalme replied:
This now panics in unit tests as the auth map can not be created without privileges. I'll leave of the above `panic()` for now, you can then fix this by transforming maps/authmap to a Hive cell.

joestringer commented on bpf/lib/utime.h:
```diff
@@ -0,0 +1,41 @@
+ * Dividing nanoseconds by 2^9 yields roughly half microsecond accuracy without
+ * any loss when converting from seconds (as 1e9/2^9 is an integer), but avoids
+ * expensive 64-bit divisions.  With this shift the range of an u64 is ~300000
+ * years instead of ~600 years if left at milliseconds.
```
```suggestion
 * Number of bits to shift a monotonic 64-bit nanosecond clock for utime unit.
 * Dividing nanoseconds by 2^9 yields roughly half microsecond accuracy without
 * any loss when converting from seconds (as 1e9/2^9 is an integer), but avoids
 * expensive 64-bit divisions.  With this shift the range of an u64 is ~300000
 * years instead of ~600 years if left at nanoseconds.
```
?

joestringer replied:
Incidentally, while the last statement is technically true, we wouldn't be able to initialize or make `utime_get_time()` calculations based on a value that represents >~600 years currently because we first start with a timestamp in nanoseconds, thereby restricting us to `EPOCH < ktime_ns < ~600y`, then apply the math to render this into UTIME.

Not really sure there's much actionable out of this observation, other than to say that the folks dealing with the "Year 2554" problem should have fun with it.

jrajahalme replied:
Fixed the comment, thanks!

While we start with a timestamp in nanoseconds, that is nanoseconds since boot, so if the system is rebooted every 600 years, utime can and will represent Unix epoch based expirations times upto year 301460 :-) Also the userspace tooling translating a given expiration `time.TIme` to `utime` units never changes the timestamp to nanoseconds only. There is also a unit test that checks that the accuracy of timestamp translations from `time.Time` to `utime` and back to `time.Time` keeps microsecond accuracy from 1970 (second 0) to year 301460 (second 944_4732_965_739) iterating a large prime and using the value both as seconds and nanoseconds to not test only whole secnods.

joestringer replied:
💡 Ah right, I don't know why I assumed it was time since boot elsewhere then switched my assumption to time since Unix Epoch here. Thanks for explaining!

joestringer commented on pkg/datapath/alignchecker/alignchecker.go:
Do we need to add utime map to the alignchecker too(?) (maybe not if it's always just going to be a one-element array with a single u64)

jrajahalme replied:
I thought that to be unnecessary for the reasons you stated here.

joestringer commented on pkg/maps/utime/utime.go:
```diff
@@ -0,0 +1,241 @@
+	// Number of bits to shift a monotonic 64-bit nanosecond clock for utime unit.  Dividing
+	// nanoseconds by 2^9 yields roughly half microsecond accuracy but avoids expensive 64-bit
+	// divisions in the datapath.  With this shift the range of an u64 is ~300000 years
+	// instead of ~600 years if left at milliseconds.
```
Similar comment to my earlier comment here.

Might also be useful to mention that this must be in sync with the datapath symbol and cannot be changed without considering the upgrade path.

jrajahalme replied:
Fixed the milli->nano at the end and added:
```
	// Must be kept in sync with `UTIME_SHIFT` in datapath (bpf/lib/utime.h), any changes to
	// this will have an upgrade impact.
```

joestringer commented on bpf/lib/maps.h:
```diff
@@ -99,6 +99,16 @@ struct {
 } AUTH_MAP __section_maps_btf;
 #endif
 
+#ifdef UTIME_MAP
```
I'm not sure the best place to document this, but I think it would be useful information for readers to see that "UTIME_MAP is a single-element array containing the boot time in UTIME format, ie nanoseconds>>9".

jrajahalme replied:
Good point, adding this here:
```
/*
 * UTIME_MAP is a single-element array containing the adjusted boot time in
 * UTIME format, ie. nanoseconds>>9. "adjusted" here means that it includes the
 * time difference between BOOTTIME and MONOTONIC clocks so that we can add this
 * value to the time given by ktime_get_ns(), which does not account for suspend
 * time. Once the oldest kernel supported is Linux 5.8, we can use
 * ktime_get_boot_ns() instead so that this value does not need to contain an
 * estimate of the difference between BOOTTIME and MONOTONIC clocks any more.
 */
```

jrajahalme replied:
Figured this is as good a place for this documentation as any other :-)

joestringer reviewed (commented):
Just minor nits left. I looked at the datapath code, but I didn't look in detail at the utime calculation details particularly around initialization.

ti-mo commented on pkg/maps/utime/utime_test.go:
```diff
@@ -0,0 +1,68 @@
+
+	// Local testing showed a difference of less than a second,
+	// fail test if more than 5 seconds
+	c.Assert(math.Abs(diff.Seconds()) < float64(5), Equals, true)
```
`c.Assert(diff < (time.Second * 5), Equals, true)`?

jrajahalme replied:
Right :-)

ti-mo commented on pkg/maps/utime/utime_test.go:
```diff
@@ -0,0 +1,68 @@
+var _ = Suite(&utimeSuite{})
+
+func (s *utimeSuite) TestUTime(c *C) {
+	c.Assert(maxSeconds, Equals, 944_4732_965_739)
```
`9_444_732_965_739`

jrajahalme replied:
Fixed

ti-mo commented on pkg/maps/utime/utime.go:
```diff
@@ -0,0 +1,244 @@
+	utimeShift = 9
+	// 10^9 has 9 trailing zeroes also in binary, so they can be shifted off without any loss
+	// of accuracy.
+	secsToUtimeMultiplier = 1000_000_000 >> utimeShift // integer value (1953125)
```
`1_000_000_000`

jrajahalme replied:
Fixed, thanks!

ti-mo commented on pkg/maps/utime/utime.go:
```diff
@@ -0,0 +1,244 @@
+	UpdateController(name string, params controller.ControllerParams)
+}
+
+// MapCreate will create an encrypt map
```
This looks like it's copy-pasted.

ldelossa replied:
Ditto, what is this? At the least a better explanation would be helpful

jrajahalme replied:
Yes it was, fixed.

ti-mo commented on pkg/maps/utime/utime.go:
```diff
@@ -0,0 +1,244 @@
+		// Add controller for keeping clock in sync for NTP time jumps and any difference
+		// between monotonic and boottime clocks.
+		ctrlName := "sync-utime"
+		controllerManager.UpdateController(ctrlName,
```
Should we try an `OpenOrCreate()` first before launching the controller?

jrajahalme replied:
the sync call right above does it. I guess you mean that we should not start the controller if `OpenOrCreate()` fails? Not sure it that makes a difference if the agent is going to exit on this error anyway?

ti-mo commented on pkg/maps/utime/utime.go:
```diff
@@ -0,0 +1,244 @@
+	return toUTime(t.Unix(), t.Nanosecond())
+}
+
+func (t UTime) Time() time.Time {
```
Unexport?

jrajahalme replied:
Unexported all these types/functions :-)

ti-mo commented on pkg/maps/utime/utime.go:
```diff
@@ -0,0 +1,244 @@
+	return UTime(secs)*secsToUtimeMultiplier + UTime(nanos)>>utimeShift
+}
+
+func TimeToUTime(t time.Time) UTime {
```
Unexport?

ti-mo commented on pkg/maps/utime/utime.go:
```diff
@@ -0,0 +1,244 @@
+
+// GetBoottime returns the kernel boot time.
+// We parse it from /proc/stat. GetBoottime() should be invoked only occasionally.
+func GetBoottime() (t time.Time, err error) {
```
Unexport?

ti-mo reviewed (commented):
Leaving a comment as I only reviewed the utime commit.

We're adding a map and allocating a pin path just for plumbing one u64 into the datapath. Has this been discussed with the group? I don't want to block this PR, but I would like to raise the point of technical debt here. This map and its types have an overly-specific name, so adding more of these values and their corresponding controllers will be problematic.

Could this utime package be refactored into a generic runtime configuration subsystem ahead of time? This way we don't have to handle renaming/removing this map later if/when it's introduced. Future additions will also be much more straightforward, as this PR already introduces much of the needed infrastructure. (just missing atomic reads from bpf side, though I'm not sure if this requires support from the kernel)

joamaki commented on pkg/maps/utime/utime.go:
```diff
@@ -0,0 +1,244 @@
+
+var (
+	once     sync.Once
+	uTimeMap UTimeMap
```
Let's please make this into a cell from the start and not add more global variables.
Ping me on slack and I walk through it in detail.

joestringer replied:
(Please capture the output of any Slack discussion or missing docs in https://docs.cilium.io/en/latest/contributing/development/hive/#id1 , I would like to make sure this information is publicly available and doesn't require any contributor to schedule a synchronous session with a committer in order to make progress)

jrajahalme replied:
@mhofstetter volunteered to transition this to a Hive Cell in an "immediate" follow-up PR. He did the same for the auth map so that would fit me perfectly :-)

joamaki replied:
awesome thanks @mhofstetter

jrajahalme replied:
Unfortunately none of the `pkg/maps` are defined as Cells yet, so there is no example to go by.

mhofstetter replied:
branch with a initial proposal is ready - will open a PR as soon as this PR is merged

mhofstetter replied:
-> https://github.com/cilium/cilium/pull/24830

ldelossa reviewed (approved):
LGTM

jrajahalme committed:
maps: Add datapath runtime config map (configmap)

Add a new config map for passing runtime configuration values that may
change often enough to not warrant recompiling the affected bpf programs,
and whose changes will not have an effect to the structure of the bpf
programs.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
utime: Add runtime config for time syncing with the datapath

Define "utime" as positive Unix epoch time, in 512 nanosecod units. This
unit is chosen as the largest exponentian of base 2 that yields a whole
integer multiplier to get from seconds to utime units without any loss of
accuracy, and to gain more future time range. "utime" is an unsigned 64
bit integer, so the range is from the Unix epoch (1.1.1970) to September
3rd of year 301261.

With this formulation, the result of ktime_get_ns() can be shifted right
by 9 bits to gain the current monotonic clock in "utime" units.

Finally, this commit adds an entry to the configmap, the offset
to add to the shifted monotonic clock to get the current time in Unix
epoch time. This offset is periodically updated by Cilium agent to
account for the boot time of the system, any NTP time jumps, and the
difference between boottime and monotonic clocks (such as suspend time).

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

mhofstetter commented on pkg/datapath/alignchecker/alignchecker.go:
```diff
@@ -99,8 +100,9 @@ func CheckStructAlignments(path string) error {
-		"auth_info":              {reflect.TypeOf(auth.AuthInfo{})},
+		"auth_key":               {reflect.TypeOf(authmap.AuthKey{})},
+		"auth_info":              {reflect.TypeOf(authmap.AuthInfo{})},
+		"config_value_0":         {reflect.TypeOf(configmap.ConfigValue0{})},
```
`config_value_0` -> `runtime_config_0` ?

mhofstetter commented on pkg/maps/configmap/config_map.go:
```diff
@@ -0,0 +1,187 @@
+type ConfigValue0 struct {
+	// 'uTimeOffset' is the number of UTime units (512ns each) to be added to
+	// the monotonic time to get the Unix epoch time.
+	uTimeOffset utime.UTime `align:"utime_offset_512"`
```
`utime_offset_512` -> `utime_offset_512_ns` ?

jrajahalme replied:
Thanks!

lmb commented on bpf/lib/maps.h:
```diff
@@ -99,6 +99,21 @@ struct {
+struct {
+	__uint(type, BPF_MAP_TYPE_ARRAY);
+	__type(key, __u32);
+	__type(value, struct runtime_config_0);
```
Using a struct here means that the smallest update you can make is the whole struct. In user space that will end up requiring a lookup of the current value, patching only the necessary parts and then writing that back. This is kind of unergonomic in terms of what code you write, and it also suffers badly if you have concurrent modifications.

What if instead of `struct runtime_config` we used `u64` here instead? That implies that the largest such runtime config is 8 bytes, but my gut feeling is that this is probably OK. To make "allocating" config easy you could add an enum.

jrajahalme replied:
`u64` may be enough, but I was not ready to make that call without discussion. Current implementation caches the values in the userspace, so entries are updated only when needed. This design also periodically pulls in the config values from the configmap package so that the layout of the config data is only ever known in the configmap pkg and the datapath `lib/common.h` definitions. Also the size on 64 bytes likely makes no perf impact compared to the system call overhead. Also on the datapath side that is a typical cache line size, and I've also understood that bpf code accesses the map data by reference. Even if copied, a single cache line is likely not significantly slower than single `u64`.

That being sade, I had to choose a desing to propose. It would be nice if the datapath folks would discuss this over :-)

lmb commented on pkg/maps/configmap/config_map.go:
```diff
@@ -0,0 +1,181 @@
+	MaxEntries = 256
+)
+
+// ConfigKey is the index to the Unix time offset. Only value 0 is supported for now.
```
Leftover reference to Unix time?

jrajahalme replied:
Yes, from separating the commits, will clean-up.

lmb commented on pkg/maps/configmap/config_map.go:
```diff
@@ -0,0 +1,181 @@
+// GetKeyPtr returns the unsafe pointer to the BPF key
+func (k *ConfigKey) GetKeyPtr() unsafe.Pointer { return unsafe.Pointer(k) }
+
+// NewValue returns a new empty instance of the structure represeting the BPF
```
Nit: `representing`.

jrajahalme replied:
Thanks!

lmb commented on bpf/lib/maps.h:
```diff
@@ -99,6 +99,21 @@ struct {
+	__type(value, struct runtime_config_0);
+	__uint(pinning, LIBBPF_PIN_BY_NAME);
+	__uint(max_entries, CONFIG_MAP_SIZE);
+} CONFIG_MAP __section_maps_btf;
```
This map should probably be RO from the BPF program side?

jrajahalme replied:
Yes. Is there a flag for that? The only reference I could find from the datapath code is this:
```
tests/pktgen.h: * with kernels before 5.2 that lack read-only map support. This test suite
```

lmb replied:
`BPF_F_RDONLY_PROG` is your friend!

lmb replied:
This didn't work out for some reason?

P.S. GitHub UI being terrible, this refers to BPF_F_RDONLY_PROG

jrajahalme replied:
We have not defined that flag in userspace, nor are we using it anywhere in the datapath. I don't know what happens on older kernels not supporting this flag if it is set, and don't feel like this PR is the right place to experiment on that. I feel like this would be a separate piece of work to intro this flag (and maybe others) on the userspace side, make use of it on all the maps that are read only from the datapath side, and deal with the kernel versions, if need be.

lmb commented on pkg/maps/configmap/config_map.go:
```diff
@@ -0,0 +1,181 @@
+// ConfigKey is the index to the Unix time offset. Only value 0 is supported for now.
+// +k8s:deepcopy-gen=true
+// +k8s:deepcopy-gen:interfaces=github.com/cilium/cilium/pkg/bpf.MapKey
+type ConfigKey struct {
```
Why are the `Config`* types exported? Seems like nothing is using it externally.

jrajahalme replied:
Likely for nothing, thanks for the tip :-)

lmb commented on pkg/maps/configmap/config_map.go:
```diff
@@ -0,0 +1,181 @@
+		// Add controller for keeping clock in sync for NTP time jumps and any difference
+		// between monotonic and boottime clocks.
+		ctrlName := "sync-config"
+		controllerManager.UpdateController(ctrlName,
```
Newbie question: how come that the configmap sets up the periodic task? Is that a convention? I kind of expected utime or similar to do that. Thinking this forward, we'd end up with a bit of custom code here for each config, which doesn't scale well.

jrajahalme replied:
That is a design question rather than a newbie one. I had no other use than `utime` for this in mind, so went with a minima interface between `configmap` and `utime`. Open for better proposals, such as the one you proposed above :-). Unless there are comments to the contrary or other opinions I'll go with that one :-)

lmb reviewed (commented):
I looked at the commit that adds the `configmap` and left a few comments. Overall, maybe it would make more sense if we had the following Go interface?

```go
package configmap

type Key struct{ value uint32 }

var BootTime = Key{ 0 }

func Init(...)

func Update(key Key, value uint64) error
```

Then adding a new config item is just adding a `var NewConfig = Key{}`. Since `Key.value` is not exported we can force contributors to add new keys in a single place. And we can use that var to find references to that config in the codebase.

joamaki commented on pkg/maps/configmap/doc.go:
```diff
@@ -0,0 +1,6 @@
+// Copyright Authors of Cilium
+
+// Package config stores runtime configuration state for the Cilium datapath.
+// +groupName=maps
```
Why is groupName here? AFAIK it only affects code generation for k8s object structs and typed clients.

jrajahalme replied:
IDK, all the packages under `pkg/maps` have this `groupName=maps`.

joamaki commented on pkg/maps/configmap/config_map.go:
```diff
@@ -0,0 +1,187 @@
+	// 'uTimeOffset' is the number of UTime units (512ns each) to be added to
+	// the monotonic time to get the Unix epoch time.
+	uTimeOffset utime.UTime `align:"utime_offset_512_ns"`
+	reserved    [7]uint64   `align:"reserved"`
```
Shouldn't this `reserved` be in `ConfigValues` ?

jrajahalme replied:
Does not matter if/when changing to a simpler array of `u64` values as proposed by Lorenz above.

joamaki commented on pkg/maps/configmap/config_map.go:
```diff
@@ -0,0 +1,187 @@
+
+// syncConfig updates the current config if needed
+func (r *ConfigMap) syncConfig(force bool) error {
+	_, err := r.OpenOrCreate()
```
Why are we calling `OpenOrCreate` again here? It was already called from `InitMap`

jrajahalme replied:
`InitMap()` called `syncConfig()` before `OpenOrCreate()`, but this is now changed.

joamaki commented on pkg/maps/configmap/config_map.go:
```diff
@@ -0,0 +1,187 @@
+
+	config := ConfigValues{
+		value0: ConfigValue0{
+			uTimeOffset: utime.GetCurrentUTimeOffset(),
```
I don't think this map should pull data from elsewhere, but rather it should be provided these from outside. E.g. `utime` should have a controller that updates the configmap. Or if it's a one-shot update then we could have `configmap` do a one-shot controller to try and set the configmap value and retry on failures.

Also consider exposing this map as `type ConfigMap interface` that has e.g. `Update(*ConfigValues)`. Makes it easier to test against and easier to modularize later.

jrajahalme replied:
Will reverse the roles, but a bit simpler with a `u64` array where different sources can "own" indices to pass runtime config values to the datapath.

joamaki commented on pkg/utime/utime.go:
```diff
@@ -0,0 +1,120 @@
+func getBoottime() (t time.Time, err error) {
+	var boottime int64
+	var delta int64
+	stat, err := os.Open(btimeInfoFilepath)
```
We're reading Linux specific files here, so this package should be under `pkg/datapath/linux`.

jrajahalme replied:
Good point, will move.

aanm commented on bpf/lib/common.h:
```diff
@@ -880,9 +880,8 @@ struct ct_entry {
 	      proxy_redirect:1, /* Connection is redirected to a proxy */
 	      dsr:1,
 	      from_l7lb:1, /* Connection is originated from an L7 LB proxy */
-	      auth_required:1,
```
shouldn't this be changed to ` reserved:1,`? Otherwise the `from_tunnel` will be in the wrong place

jrajahalme replied:
Changed to `reserved1:1,` to avoid moving `from_tunnel` around.

aanm commented on bpf/lib/common.h:
```diff
@@ -1069,9 +1068,8 @@ struct ct_state {
 	      syn:1,
 	      proxy_redirect:1,	/* Connection is redirected to a proxy */
 	      from_l7lb:1,	/* Connection is originated from an L7 LB proxy */
-	      auth_required:1,
```
shouldn't this be changed to ` reserved:1,`? Otherwise the `from_tunnel` will be in the wrong place

jrajahalme replied:
there is no release with from_tunnel yet so I thought that would not matter?

aanm replied:
I'll defer to @YutaroHayakawa then.

jrajahalme replied:
Changed to `reserved1:1,` to avoid moving `from_tunnel` around.

lmb reviewed (approved):
configmap changes LGTM! Thanks for reworking :)

jrajahalme reviewed (commented):
Updated to not move `from_tunnel` bit in conntrack flags around.
